### ISSUE TITLE
I was going to clean it up, I promise!

### ISSUE THREAD

**Issue by LinusU** [sindresorhus/loud-rejection#1]
I don't know if there is much we can do about this, but it would be very cool if this didn't throw an error.

```javascript
var loudRejection = require('loud-rejection')

loudRejection()

var myp

setTimeout(function () {
  myp = new Promise(function (resolve, reject) {
    setTimeout(reject, 100, new Error('Silence me'))
  })
}, 100)

setTimeout(function () {
  myp.catch(function (err) {
    console.log('Catched!')
  })
}, 300)
```

I think that it has to go into the `Promise` implementation thought, and I guess that that ship has sailed.

My idea was to do this when the promise is thrown away by the garbage collector, if it hasn't been handled by then it won't ever be. But maybe this has problems as well?

**Comment by sindresorhus** [comment]
Hmm, since this is for CLI tools, I could defer the console.error and exiting until `process.on('exit')`.

See the docs: https://nodejs.org/api/process.html#process_event_unhandledrejection

Could probably do something like the example in https://nodejs.org/api/process.html#process_event_rejectionhandled and keep track of unhandled rejection that are not eventually caught.

**Comment by LinusU** [comment]
Hmm, yes that could possible work actually. Nice find!

**Comment by sindresorhus** [comment]
Pseudo code:

```js
var unhandledRejections = [];
process.on('unhandledRejection', function(reason, p) {
    unhandledRejections.push(p);
});
process.on('rejectionHandled', function(p) {
    var index = unhandledRejections.indexOf(p);
    unhandledRejections.splice(index, 1);
});

process.on('exit', function () {
  if (unhandledRejections.length > 0) {
    unhandledRejections.forEach(function (x) {
      console.error(x.reason);
    });
    process.exitCode = 1;
  }
});
```

**Comment by sindresorhus** [comment]
Wanna do a PR?

**Comment by LinusU** [comment]
Sure! Might take a day thought :rocket:

**Comment by sindresorhus** [comment]
@LinusU Any progress? No worries if not. Just asking as I'd like to include this in [AVA](https://github.com/sindresorhus/ava/issues/80), but it's blocked by this issue.

**Comment by LinusU** [comment]
None yet, I'm on the airport going home now so if I'm not too tired I'll fix it tonight...

**Comment by jamestalmage** [comment]
Closed via #5
